include/swift/AST/AccessScope.h

```diff
@@ -46,25 +59,45 @@ class AccessScope {
   bool operator==(AccessScope RHS) const { return Value == RHS.Value; }
   bool operator!=(AccessScope RHS) const { return !(*this == RHS); }
   bool hasEqualDeclContextWith(AccessScope RHS) const {
+    if (isPublic())
+      return RHS.isPublic();
+    if (isPackage())
+      return RHS.isPackage();
     return getDeclContext() == RHS.getDeclContext();
   }
 
-  bool isPublic() const { return !Value.getPointer(); }
-  bool isPrivate() const { return Value.getPointer() && Value.getInt(); }
+  bool isPublic() const {
+    return !Value.getPointer() && Value.getInt() == AccessLimitKind::None;
+  }
+  bool isPrivate() const {
+    return Value.getPointer() && Value.getInt() == AccessLimitKind::Private;
+  }
   bool isFileScope() const;
   bool isInternal() const;
+  bool isPackage() const {
+    return !Value.getPointer() && Value.getInt() == AccessLimitKind::Package;
+  }
 
   /// Returns true if this is a child scope of the specified other access scope.
-  ///
+  /// The 'package' scope is between internal and public.
   /// \see DeclContext::isChildContextOf
   bool isChildOf(AccessScope AS) const {
-    if (!isPublic() && !AS.isPublic())
-      return allowsPrivateAccess(getDeclContext(), AS.getDeclContext());
-    if (isPublic() && AS.isPublic())
-      return false;
-    return AS.isPublic();
+    if (isInternalOrLess()) {
+      if (AS.isInternalOrLess())
+        return allowsPrivateAccess(getDeclContext(), AS.getDeclContext());
+      else
+        return AS.isPackage() || AS.isPublic();
+    }
+    if (isPackage())
+      return AS.isPublic();
+
+    // If this is public, it can't be less than access level of AS, so return
+    // false
+    return false;
```

Comment:
Hang on a minute, I have some questions.

The old version ended with:

```cpp
    if (isPublic() && AS.isPublic())
       return false;
     return AS.isPublic();
```

Does the new version still return `false` if both `this` and `AS` are public? Does it return `false` if `this` is public but `AS` is less than package? If not, why is that behavior change correct?

Now the harder question: Do we actually want `package` to be treated as a child of `public` here, or should they be treated as synonyms for the purpose of `isChildOf()`? I don't have a great sense for how `isChildOf()` is used, but I don't think it's just an "is this scope more restrictive than that scope" thing--I think it's more about whether one scope encloses the other.

Reply:
Searching callsites shows it's mostly used to compute the min access scope, e.g. 
```    
auto minAccessScope = AccessScope::getPublic();
if (typeAccessScope.isChildOf(minAccessScope) { ... }
```
I was going to rename it as `isLessThan()` or `isMoreRestrictiveThan()`  @xedin  any thoughts?

Reply:
> Does the new version still return false if both this and AS are public? Does it return false if this is public but AS is less than package? If not, why is that behavior change correct?

Yes it returns false for both. 
```
     if (isPublic() && AS.isPublic())
       return false;
     return AS.isPublic();
```
can be translated to 
```
     if (isPublic()) {
       if (AS.isPublic()) 
         return false;
       else
         return AS.isPublic(); // false
    }
     return AS.isPublic(); // return false since all possibilities are computed above 
```

Reply:
The call sites of isChildOf(arg) seems to be using it to calculate min access level, e.g. 

```
  auto minAccessScope = AccessScope::getPublic();
  if (typeAccessScope.isChildOf(minAccessScope)
``` 
so it appears to be calculating whether "this is more restrictive than the arg scope". Since `package` is more restrictive, e.g. a public func shouldn't have a package return type, it should be considered a child of public arg.  Maybe it'd be worth renaming this func to `isLessThan(arg)` or `isMoreRestrictiveThan(arg)`? cc @xedin @xymus any thoughts?